### PR TITLE
feat(platform/blocks): Add table input UI and builder block

### PR DESCRIPTION
**Pull request by ntindle**
https://github.com/user-attachments/assets/909a6ecf-5731-424c-8dee-fe25db907365

### Need 💡

This PR introduces a new "Table Input" block and corresponding UI component, allowing users to easily input structured, tabular data directly within the agent builder. This addresses the need for a user-friendly way to define custom column headers and populate rows of data, which is then output as a list of dictionaries.

### Changes 🏗️

*   **New `TableInputBlock` (backend):** A new block (`backend/backend/blocks/table_input.py`) has been added. It defines an `Input` schema with `headers` (a list of strings for column names) and `value` (a list of dictionaries representing table rows). The block outputs the `value` data in the specified dictionary format.
*   **New `NodeTableInput` Component (frontend):** A new React component (`frontend/src/components/node-table-input.tsx`) was created to render an editable table UI, supporting dynamic row addition/removal and cell editing.
*   **Frontend Integration:**
    *   `NodeGenericInputField` and `NodeObjectInputTree` were updated to pass `parentContext` down the component hierarchy.
    *   `NodeArrayInput` was modified to conditionally render the new `NodeTableInput` component. It now detects when an array field (`selfKey` is "value") is part of a parent context that defines `headers`, indicating it should be rendered as a table.

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] Add a "Table Input" block to the builder.
  - [x] Define custom headers (e.g., "Name", "Email").
  - [x] Add several rows of data using the table UI.
  - [x] Verify that adding, editing, and removing rows works as expected.
  - [x] Connect the output of the "Table Input" block to another block (e.g., a "Print" block) and confirm the output format is a list of dictionaries with the defined headers as keys.
  - [x] Test with an empty table (no rows).
  - [x] Test with no headers defined (should default).
  - [x] Test that an empty row returns empty data (is this a good behavior?

example output of the block
```
{
  "advanced": false,
  "column_headers": [
    "Col 1",
    "Col 2",
    "Col 3"
  ],
  "name": "table_input",
  "value": [
    {
      "Col 1": "row 1",
      "Col 2": "row 1",
      "Col 3": "row 1"
    },
    {
      "Col 1": "val 1",
      "Col 2": "val 2",
      "Col 3": "val 3"
    }
  ]
}
```

---
<a href="https://cursor.com/background-agent?bcId=bc-b8d31867-1034-4374-852c-b92ca69cc399">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-cursor-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-cursor-light.svg">
    Open in Cursor
  </picture>
</a>
<a href="https://cursor.com/agents?id=bc-b8d31867-1034-4374-852c-b92ca69cc399">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-web-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-web-light.svg">
    Open in Web
  </picture>
</a>

### CODE REVIEW

**Comment by AutoGPT-Agent** [comment]
Thank you for this well-documented PR that adds a new Table Input block and UI component! The implementation looks solid, with both backend and frontend changes properly coordinated.

The PR has a clear description outlining the need and detailed changes. Your test plan is comprehensive, showing you've thoroughly tested the functionality.

One minor issue to address:

- The PR title doesn't follow our conventional commit format. It should include a type prefix and scope, such as: `feat(blocks): Add table input ui and builder block` or `feat(platform/blocks): Add table input ui and builder block`

Once you update the PR title to follow the conventional commit format, this PR will be ready to merge. Great job on implementing this useful feature!

**Comment by ntindle** [comment]
@claude whats goin on with this guy and what do we need to do to get it fucntional and high quality

**Comment by AutoGPT-Agent** [comment]
Thanks for this well-structured PR! The table input functionality will be a valuable addition to the platform.

The implementation looks solid with comprehensive tests for the backend and a clean UI component for the frontend. I appreciate the detailed description and thorough test plan you've included.

One minor issue needs to be addressed before merging:

- The PR title doesn't follow our conventional commit format. Please update it to include the type and scope, for example: `feat(platform/blocks): Add table input ui and builder block`

Otherwise, this is ready to go! The changes are well-scoped, focused on the task at hand, and the code quality looks good.

**Comment by AutoGPT-Agent** [comment]
Thank you for this well-documented PR that adds a new Table Input block and corresponding UI component! The implementation looks solid, with appropriate tests and a clean UI implementation.

The PR is almost ready to merge, but there's one minor issue to address:

**PR Title Format**: The current title "Add table input ui and builder block" doesn't follow our conventional commit format. Please update it to include the appropriate type and scope, for example: `feat(platform/blocks): add table input UI and builder block`

Otherwise, your implementation is thorough with:
- A well-structured backend implementation with appropriate tests
- A clean frontend implementation that extends existing components
- A complete test plan that you've verified

Once you update the title format, this should be ready to merge!

**Comment by AutoGPT-Agent** [comment]
Thank you for this well-structured PR that adds a Table Input block and its corresponding UI components!

The implementation looks solid with comprehensive test coverage, and your PR description clearly explains both the need for this feature and the specific changes made.

### Primary Issue

Your PR title needs to follow our conventional commit format. Please update it to include the type and scope, for example:

```
feat(platform/blocks): Add table input ui and builder block
```

### Minor Notes

- The implementation looks thorough with good test coverage
- I appreciate the comprehensive checklist that's been completely filled out
- Your code changes are focused on the stated feature without unrelated modifications

Once you update the PR title to follow our conventional commit format, this PR should be ready for approval.

**Comment by AutoGPT-Agent** [comment]
Thank you for submitting this comprehensive PR! The Table Input UI component and builder block look well-implemented, and your test plan is thorough.

Your PR description clearly explains the need for this feature and the changes you've made. The code changes look good, with appropriate tests for the backend functionality and a clean implementation of the table UI component.

There's just one small issue to address before we can merge this PR:

- The PR title needs to follow our conventional commit format. Please update it to include both the type and scope, like: `feat(platform/blocks): add table input ui and builder block`

Otherwise, this is a well-structured PR that adds a useful new feature to the platform. Once you update the title, it should be good to go!

**Comment by AutoGPT-Agent** [comment]
Thank you for this comprehensive PR that adds the Table Input block and UI component. The implementation looks thorough with both frontend and backend components properly integrated.

### Feedback

The PR is almost ready to merge, but there's one issue that needs to be addressed:

- The PR title doesn't follow our conventional commit format. It should start with a type (like "feat:") and include a scope. Please update the title to something like: "feat(blocks): add table input ui and builder block"

Otherwise, this is excellent work:

- The PR description clearly explains the need and changes
- The test plan is thorough and all items are checked
- The implementation includes both backend and frontend components
- Tests are included for the backend functionality
- The UI components look well-designed for both the builder and runtime environments

Once you update the PR title to follow the conventional commit format, this will be ready to merge.

**Comment by AutoGPT-Agent** [comment]
Thank you for your contribution! The Table Input block and UI component look well-implemented and thoroughly tested.

The PR is generally in good shape, but there's one issue that needs to be addressed before it can be merged:

### PR Title Format

The PR title needs to follow our conventional commit format, which requires a type and scope. Please update your PR title to something like:

```
feat(blocks): add table input UI and builder block
```

Or if you prefer to be more specific about the scope:

```
feat(platform/blocks): add table input UI and builder block
```

Your implementation looks solid, with comprehensive tests and good coverage of edge cases. I particularly appreciate:

1. The thorough test plan in your PR description
2. The well-organized UI component that handles various states
3. The comprehensive test suite for the backend component

Once you update the PR title to follow our conventional commit format, this PR should be ready to merge.

**Comment by AutoGPT-Agent** [comment]
Thank you for this well-structured PR that adds table input functionality to the platform! The implementation looks solid, with comprehensive backend testing and a clean UI component for table manipulation.

A few observations:

1. **PR Title Format**: Your PR title should follow our conventional commit format with a type and scope. For example: `feat(platform/blocks): Add table input UI and builder block`

2. **Implementation Quality**: I appreciate the thorough test coverage for the backend block and the clean UI implementation. The ability to define custom headers and populate rows dynamically will be very useful.

3. **Documentation**: The PR description clearly explains the need and changes, which makes it easy to understand the purpose of this feature.

Before merging, please update the PR title to follow the conventional commit format with an appropriate type (likely `feat`) and scope. Otherwise, the implementation looks ready to go!

**Review by ograce1421 — DISMISSED** [review]

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx [hunk 1]
```diff
@@ -211,6 +221,101 @@ export function RunAgentInputs({
...
+          <Button
+            type="button"
+            variant="outline"
+            size="sm"
+            onClick={addRow}
+            className="w-full"
+          >
+            <Plus className="mr-2 h-4 w-4" weight="bold" />
+            Add Row
+          </Button>
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx:308]: Let's use `<Button variant="outline">` from atoms [from atoms](https://github.com/Significant-Gravitas/AutoGPT/tree/dev/autogpt_platform/frontend/src/components/atoms/Button) instead here. I have updated the location of the files, so if you rebase, now:
```
- "@/components/ui/button";
+ "@/components/__legacy__/ui/button";
```
so is more obvious we shouldn't be using these ones 👍🏽

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx [hunk 2]
```diff
@@ -211,6 +221,101 @@ export function RunAgentInputs({
...
+        </div>
+      );
+      break;
+    }
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx:312]: Shouldn't we use here `node-table-input` ?

Comment (ntindle) [reply autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx:312]: The `node-table-input` is builder-specific functionality for configuring the table. I've added comments to make that clearer :)

**Comment by Pwuts** [comment]
Does pasting into the table work?

File: autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts [hunk 3]
```diff
@@ -114,6 +116,20 @@ export type BlockIOArraySubSchema = BlockIOSubSchemaMeta & {
...
+export type TableRow = Record<string, TableCellValue>;
+
+export type BlockIOTableSubSchema = BlockIOSubSchemaMeta & {
+  type: "array";
+  format: "table";
+  items: BlockIOObjectSubSchema;
+  const?: TableRow[];
+  default?: TableRow[];
+  secret?: boolean;
+};
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts:131]: Why do we need an out-of-standard schema for this? Shouldn't any array of objects with simple values be rendered as a table?

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx [hunk 4]
```diff
@@ -43,7 +46,14 @@ export function RunAgentInputs({
 }: Props & React.HTMLAttributes<HTMLElement>) {
   const { handleUploadFile, uploadProgress } = useRunAgentInputs();
 
-  const dataType = determineDataType(schema);
+  let dataType = determineDataType(schema);
+
+  // Override dataType if format is "table"
+  // This is needed because determineDataType doesn't always see the format field
+  if ("format" in schema && (schema as any).format === "table") {
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx]: looks like a bad cast

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx [hunk 5]
```diff
@@ -43,7 +46,14 @@ export function RunAgentInputs({
...
+    dataType = DataType.TABLE;
+  }
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx]: why not amend `determineDataType` instead?

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx [hunk 6]
```diff
@@ -211,6 +221,101 @@ export function RunAgentInputs({
...
+                  <th className="w-10 px-3 py-2"></th>
+                </tr>
+              </thead>
+              <tbody>
+                {tableData.map((row, rowIndex) => (
+                  <tr key={rowIndex} className="border-t dark:border-gray-700">
+                    {headers.map((header) => (
+                      <td key={header} className="px-3 py-1">
+                        <input
+                          type="text"
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx:273]: Follow-up: make this support number, date, and boolean input?

File: autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts (hunk 3 above)
Comment (ntindle) [reply autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts:131]: No there’s no constraint that an array of objects match in type

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx (hunk 6 above)
Comment (ntindle) [reply autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx:273]: Yes that’s the plan

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx (hunk 5 above)
Comment (ntindle) [reply autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx]: lol good point

File: autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx (hunk 4 above)
Comment (ntindle) [reply autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx]: good spot

Comment (ntindle) [reply autogpt_platform/frontend/src/app/(platform)/library/agents/[id]/components/AgentRunsView/components/RunAgentInputs/RunAgentInputs.tsx]: i missed updating schema with the correct format type

**Comment by ntindle** [comment]
> Does pasting into the table work?

yes

**Comment by Pwuts** [comment]
@coderabbitai review this

File: autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts (hunk 3 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts:131]: Isn't that what the `schema.items` does? Specify the type for every single item in the array? And if `items` is an object schema, that means it can be rendered as a table.
